### ISSUE TITLE
Responsive Images

### ISSUE THREAD

**Issue by Mischback** [Mischback/mischback.de#31]
Though I have already created a tool to generate the required image variants to make responsive images work ([ImP](https://github.com/Mischback/imp)), this might not be sufficient / desired for this project.

## Requirements

- [x] determine required image sizes
  - [x] 320px
  - [x] 480px
  - [x] 640px
  - [x] 960px
  - [x] 1280px
  - [x] 1600px
  - [x] 1920px
- ~~integrate in the overall build process (pre-processing, Sphinx processing, post-processing)~~
- ~~skip image processing during (local) development~~
- [x] determine file formats
  - [x] PNG (-> **default** and *fallback* format)
  - [x] WebP(?)
  - [x] Avif(?)
- ~~control which image sizes / file formats will be generated by using a configuration file *(or include this configuration in one of the existing configuration files* ``conf.py`` *or* ``pyproject.toml`` *)*~~
- [x] rely on [libvips](https://github.com/libvips/libvips) / [pyvips](https://github.com/libvips/pyvips)
  - https://libvips.github.io/pyvips/index.html 
- [x] integrate into the layout
    - [x] *Images* still should have horizontal margins in mobile layout, see admonitions for reference!

**Comment by Mischback** [comment]
## Resources

- [MDN: Responsive Images](https://developer.mozilla.org/en-US/docs/Learn/HTML/Multimedia_and_embedding/Responsive_images)
- [Google Guide having details of the difference between ``<img>`` with ``sizes`` and ``srcset`` vs. ``<picture>`` with ``<source>``](https://web.dev/learn/design/picture-element/)
  - one of the examples include the use case to include *one file* in different ``<source>`` elements *hinting different pixel densities*
  - this might be really useful:
    - Create all images as multiplies of the smallest dimensions
    - include the larger ones in all ``<source>`` elements of the smaller ones
    - **Needs heavy testing**
- [Quite detailled guide, covering ``img`` aswell as ``picture``, file formats and art direction](https://webdesign.tutsplus.com/tutorials/quick-tip-how-to-use-html5-picture-for-responsive-images--cms-21015)
- [(GERMAN) Including an example that covers resolution **and** format switching](https://kulturbanause.de/blog/responsive-images-srcset-sizes-adaptive/)
- [SmashingMagazine, covering different use cases](https://www.smashingmagazine.com/2014/05/responsive-images-done-right-guide-picture-srcset/)
- [Overview of CSSTricks, including other references](https://css-tricks.com/a-guide-to-the-responsive-images-syntax-in-html/)
- [Great introduction of ``sizes`` and ``srcset`` for ``img``](https://cloudfour.com/thinks/responsive-images-the-simple-way/)
- [CLS in combination with images](https://www.smashingmagazine.com/2021/06/how-to-fix-cumulative-layout-shift-issues/#set-width-and-heights-on-images-and-iframes)
- [Responsive Images and LCP](https://www.smashingmagazine.com/2023/01/optimizing-image-element-lcp/)
- [AVIF/WEBP vs JPEG/PNG](https://www.smashingmagazine.com/2021/09/modern-image-formats-avif-webp/)
- [Evaluate the quality of compressed images, including link to relevant Python library](https://medium.com/@thusharabandara/measure-the-compression-performance-of-an-image-compression-algorithm-ea68c1839ec6)
- [Includes code examples of SSIM](https://stackoverflow.com/questions/11541154/checking-images-for-similarity-with-opencv?noredirect=1&lq=1)

### libvips

- [Save image as ``webp``](https://www.libvips.org/API/current/VipsForeignSave.html#vips-webpsave)
- [save image as ``avif``](https://www.libvips.org/API/current/VipsForeignSave.html#vips-heifsave)
- [resize image](https://www.libvips.org/API/current/libvips-resample.html#vips-resize)
  - *resizing* works by providing a factor ``scale``, which might be calculated as ``target_width/source_width``
- [faster alternative for resizing](https://www.libvips.org/API/current/libvips-resample.html#vips-thumbnail)
- [``vipsthumbnail`` cmd line interface](https://www.libvips.org/API/current/Using-vipsthumbnail.html)
- [strip meta data programatically](https://github.com/libvips/libvips/issues/1026)

### Sphinx

- https://www.sphinx-doc.org/en/master/usage/restructuredtext/basics.html#images
- https://docutils.sourceforge.io/docs/ref/rst/directives.html#image
- https://www.sphinx-doc.org/en/master/extdev/index.html#dev-extensions
- https://www.sphinx-doc.org/en/master/development/tutorials/todo.html
- [Image Collector](https://github.com/sphinx-doc/sphinx/blob/14a9289d780240bbce78ad3640e8e1b1b12df43f/sphinx/environment/collectors/asset.py#L26)
  - assuming this does the magic for automatically copying the images to the build directory
- [docutils' guide on creating directives](https://docutils.sourceforge.io/docs/howto/rst-directives.html)
- [docutils' ``Image`` source](https://github.com/docutils/docutils/blob/ae4d18314a821e61a24dc0e4f29a691b7c3b656e/docutils/docutils/parsers/rst/directives/images.py)
- [docutils' ``visit_image()`` source](https://github.com/docutils/docutils/blob/173189b4c1c095a43c9388f4edd9bf1ff5d5b49d/docutils/docutils/writers/_html_base.py#L1008)
- [Sphinx Application API](https://www.sphinx-doc.org/en/master/extdev/appapi.html#sphinx.application.Sphinx.add_node)
- [Sphinx Photofinish source](https://github.com/wpilibsuite/sphinxext-photofinish/blob/1cefcf53bca5e142971618084724b880ba59ed63/sphinxext/photofinish/__init__.py)

### Implementation

- the ``ImageDirective`` is not implemented / extended by Sphinx. It relies on the actual [implementation in docutils](https://github.com/docutils/docutils/blob/ae4d18314a821e61a24dc0e4f29a691b7c3b656e/docutils/docutils/parsers/rst/directives/images.py#L30) ([corresponding docs](https://docutils.sourceforge.io/docs/ref/rst/directives.html#image))
- the corresponding ``node`` class in ``docutils`` is basically empty; it just defines a method ``astext()``
- ``visit_image()``
  - Sphinx's ``HTML5Writer`` ([see here](https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/writers/html5.py#L564))
  - docutils' ``Writer`` ([``visit_image()``](https://github.com/docutils/docutils/blob/ae4d18314a821e61a24dc0e4f29a691b7c3b656e/docutils/docutils/writers/_html_base.py#L992))
- a custom implementation of ``EnvironmentCollector`` might be required to mark the image source files to be included into the build (see [ImageCollector implementation](https://github.com/sphinx-doc/sphinx/blob/ff852bc7c31f48e66100e4647749fc199d92ca79/sphinx/environment/collectors/asset.py#L26))

**Comment by Mischback** [comment]
## Implementation Idea

- [x] Seperate image generation from markup
  - [x] markup is handled by a dedicated Sphinx directive
    - something like ``:graphic: filename.png`` generates the whole ``<picture>`` markup, including different image resolutions and filetypes
    - ``filename.png`` might not even exist! But the *generated* **actual filenames** must then be included in the overall build process and their existence must be checked!
    - the directive must support specification of ``alt``/``title`` attributes 
  - [x] images are generated when they are included into the repository (see **generation script** below)
- [x] image-processing script / generation script
  - [x] provide mode to just generate compressed versions of a source file (e.g. a file ``sample-in.jpg`` will be stored as ``sample-out.jpg``, ``sample-out.webp`` and ``sample-out.avif``)
    - [x] **must support** *lossy* and *lossless* compression, **should** be selectable but provide a sane default based on input file (``jpg`` -> *lossy*, ``png`` -> *lossless*)
    - [x] applies a default compression level per format, **must** be configurable
    - [x] **can** determine deviation from the input file and adjust compression level dynamically
    - [x] **must** strip all meta data from the files (e.g. EXIF, ...)
  - [x] provide mode to make a source file *responsive* (e.g. a file ``sample-in.jpg`` will be processed to produce ``sample-out_large.jpg``, ``sample-out_big.jpg``, ..., ``sample-out_tiny.jpg``, ``sample-out_large.webp``, ..., ``sample-out_tiny.webp``, ``sample-out_large.avif``, ..., and ``sample-out_tiny.avif``)
    - [x] **must** the generated sizes are dependent on *widths* and are attached to a semantic name (like ``tiny``) and are configurable
    - [x] **must** be DRY: the *resized* images are then handed over to the compression mode (see above) for compression and storage

**Comment by Mischback** [comment]
## Use responsive images while mitigating CLS

See [this example @ MDN](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/source#picture_with_height_width_attributes_example) and [this recommendation by Google](https://web.dev/patterns/web-vitals-patterns/images/responsive-images/)

This might lead to rather complex HTML code, as all images are provided in 

- a) different sizes
- b) different formats

Consequently, the following code might be generated:

```html
<picture>
  <source srcset="foo-big.avif 1024w, foo-large.avif 2048w" media="(min-width: 1024px)" width="1024" height="512" type="image/avif">
  <source srcset="foo-big.webp 1024w, foo-large.webp 2048w" media="(min-width: 1024px)" width="1024" height="512" type="image/webp">
  <source srcset="foo-big.jpg 1024w, foo-large.jpg 2048w" media="(min-width: 1024px)" width="1024" height="512">

  <source srcset="foo-medium.avif" media="(min-width: 800px)" width="512" height="256" type="image/avif">
  <source srcset="foo-medium.webp" media="(min-width: 800px)" width="512" height="256" type="image/webp">
  <source srcset="foo-medium.jpg" media="(min-width: 800px)" width="512" height="256">

  <source srcset="foo-small.avif" width="256" height="128" type="image/avif">
  <source srcset="foo-small.webp" width="256" height="128" type="image/webp">

  <img src="foo-small.jpg" alt="Some description" width="256" height="128">
</picture>
```

(The source code basically replicates Google's recommendation and extends it using different image (compression) formats. This does result in rather bloated source code, as basically all images are provided in their own ``<source>`` element. On the other hand this directly enables *art direction*, assuming that changed image dimensions are dependent on the *semantic size* of the image.)

Taking [this other Google guide](https://web.dev/learn/design/picture-element/) into consideration, the code may be modified like this:

```html
<picture>
  <source srcset="foo-big.avif 1024w, foo-large.avif 2048w" media="(min-width: 1024px)" width="1024" height="512" type="image/avif">
  <source srcset="foo-big.webp 1024w, foo-large.webp 2048w" media="(min-width: 1024px)" width="1024" height="512" type="image/webp">
  <source srcset="foo-big.jpg 1024w, foo-large.jpg 2048w" media="(min-width: 1024px)" width="1024" height="512">

  <source srcset="foo-medium.avif 512w, foo-big.avif 1024w, foo-large.avif 2048w" media="(min-width: 800px)" width="512" height="256" type="image/avif">
  <source srcset="foo-medium.webp 512w, foo-big.webp 1024w, foo-large.webp 2048w" media="(min-width: 800px)" width="512" height="256" type="image/webp">
  <source srcset="foo-medium.jpg 512w, foo-big.jpg 1024w, foo-large.jpg 2048w" media="(min-width: 800px)" width="512" height="256">

  <source srcset="foo-small.avif 256w, foo-medium.avif 512w, foo-big.avif 1024w, foo-large.avif 2048w" width="256" height="128" type="image/avif">
  <source srcset="foo-small.webp 256w, foo-medium.webp 512w, foo-big.webp 1024w, foo-large.webp 2048w" width="256" height="128" type="image/webp">
  <source srcset="foo-small.jpg 256w, foo-medium.jpg 512w, foo-big.jpg 1024w, foo-large.jpg 2048w" width="256" height="128">

  <img src="foo-small.jpg" alt="Some description" width="256" height="128">
</picture>
```

This implementation **sacrifices** the *art direction* feature, but allows serving suitable images for higher DPRs.

Question is: Is this relevant? If a user surfs the website with an iPhone (DPR > 1), is he really willing to download a significantly larger image file?

**Comment by Mischback** [comment]
- https://github.com/sphinx-doc/sphinx/blob/ff852bc7c31f48e66100e4647749fc199d92ca79/sphinx/environment/collectors/asset.py#L26
- https://github.com/sphinx-doc/sphinx/blob/ff852bc7c31f48e66100e4647749fc199d92ca79/sphinx/environment/collectors/__init__.py#L15
- https://github.com/sphinx-doc/sphinx/blob/ff852bc7c31f48e66100e4647749fc199d92ca79/sphinx/environment/__init__.py#L26
- https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/application.py#L335

**Comment by Mischback** [comment]
- **config-inited**
  - emitted in ``Sphinx.__init__()`` ([sphinx/application.py #247](https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/application.py#L247))
  - the configuration is read and available
- **builder-inited**
  - emitted in ``Sphinx._init_builder()`` ([sphinx/application.py #331](https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/application.py#L331))
  - the builder is ready, meaning it *should be* completly set up and good to go
  - the ``DirHTMLBuilder`` does not do stuff on its own, it relies on the implementation of ([``StandaloneHTMLBuilder``](https://github.com/sphinx-doc/sphinx/blob/ba080286b06cb9e0cadec59a6cf1f96aa11aef5a/sphinx/builders/html/__init__.py#L188))
    - at this point there is a ``self.imagedir`` attribute available on the ``builder`` (the ``builder`` is available as ``app.builder``)
- **env-get-outdated**
  - emitted *after* Sphinx has determined ``added``, ``changed`` and/or ``removed`` documents
- **env-before-read-docs**
  - emitted *just before* Sphinx (re-) reads documents
- **env-purge-doc**
  - emitted for *every document* that is (re-) read
  - (should) clean all traces of the document from the build environment
- **source-read**
  - emitted for *every document* after it is (re-) read
- **doctree-read**
  - emitted for *every document*
  - emitted when the doctree is read and parsed, before it is *pickled*
- **env-merge-info**
  - *optional*, if parallel reading is enabled
  - required/utilized to merge the different environments
- **env-updated**
  - emitted when **all** doctrees are up to date
- **env-check-consistency**
  - *only* **experimental**
- **doctree-resolved**
  - emitted for *every document*
  - document is ready to be passed to the writer

**Comment by Mischback** [comment]
```Python
"""Generate the required (HTML) markup for responsive images.

This extension handles the markup generation only, the actual images are
generated using the utility script in ``util/process-image.py``.
"""


# Python imports
from functools import total_ordering
from pathlib import Path

# Sphinx imports
from sphinx.environment.collectors import EnvironmentCollector
from sphinx.util import FilenameUniqDict, logging
from sphinx.util.docutils import SphinxDirective
from sphinx.util.i18n import search_image_for_language

# external imports
# ``imagesize`` is as of now a Sphinx dependency already and *should* be
# available.
import imagesize
from docutils import nodes
from docutils.parsers.rst import directives, roles

# get a module-level logger
# Note: This logger is derived from Sphinx, which has some (desired)
#       implications:
#         - emitting warnings adds up to the overall count of warnings during
#           a build. When running with ``-W``, this will fail the build!
#         - no direct control over the logging settings! This is slightly
#           inconvenient during development and mitigated by using the good ol'
#           ``print()`` statement.
logger = logging.getLogger(__name__)

# Map filename extensions to mime types.
EXT_TO_MIME = {
    ".jpg": "image/jpeg",
    ".jpeg": "image/jpeg",
    ".png": "image/png",
    ".webp": "image/webp",
    ".avif": "image/avif",
}


def _get_image_size(filename):
    """Return the image's width and height.

    This wraps around ``imagesize`` to make its interface usable by the
    extension's code.

    There is a similar implementation in Sphinx's codebase. This function
    removes all exception handling, as exceptions should be handled in the
    calling function/method.

    Parameters
    ----------
    filename : str or Path
        The filename including the path relative to Sphinx's working directory.

    Returns
    -------
    tuple, None
    """
    size = imagesize.get(filename)

    if size[0] == -1:
        return None
    elif isinstance(size[0], float) or isinstance(size[1], float):
        return (int(size[0]), int(size[1]))

    return size


@total_ordering
class ImageCandidate:
    """Represent a single image source file.

    When an image source file is represented as an instance of this class, it
    is ensured that
    a) the file is available and readable
    b) required meta informations are available, including the image's
       dimensions, its filetype/MIME type and its source path

    Parameters
    ----------
    filename : str or Path
        The filename including the path relative to Sphinx's working directory.

    Attributes
    ----------
    src_path : Path
        The actual path (relative to Sphinx's working directory) to the image
        file.

        Note: This is **not** the final path for Sphinx's output/build process.
    width: int
        The actual width of the image source file, provided in **px** (but
        without the actual unit).
    height: int
        The actual height of the image source file, provided in **px** (but
        without the actual unit).
    """

    class ProcessingError(RuntimeError):
        """Indicate problems during processing."""

    def __init__(self, filename, file_width=None, file_height=None):
        file_dim = _get_image_size(filename)

        if file_dim is None:
            if (file_width is None) or (file_height is None):
                raise self.ProcessingError("Could not determine image dimensions")
            else:
                file_dim = (file_width, file_height)

        self.src_path = filename
        self.width = file_dim[0]
        self.height = file_dim[1]
        self.mime = EXT_TO_MIME[filename.suffix]

    def __eq__(self, other):
        """Check equality with ``other`` object."""
        # see https://stackoverflow.com/a/2909119
        # see https://stackoverflow.com/a/8796908
        if isinstance(other, ImageCandidate):
            return self.__key() == other.__key()
        return NotImplemented

    def __hash__(self):
        """Provide a unique representation of the instance."""
        # see https://stackoverflow.com/a/2909119
        return hash(self.__key())

    def __lt__(self, other):
        """Check equality with ``other`` object."""
        # see https://stackoverflow.com/a/8796908
        if isinstance(other, ImageCandidate):
            return self.__key() < other.__key()
        return NotImplemented

    def __repr__(self):
        """Provide an instance's ``representation``."""
        # see https://stackoverflow.com/a/12448200
        return "<ImageCandidate(filename={}, file_width={}, file_height={})>".format(
            self.src_path.__repr__(), self.width.__repr__(), self.height.__repr__()
        )

    def __key(self):
        """Provide internal representation of the instance."""
        # see https://stackoverflow.com/a/2909119
        return (self.src_path, self.width, self.height, self.mime)


class ImageSources:
    """Collector for all corresponding sources, represented as ``ImageCandidate``.

    Internally this class works on a Python ``set``. The actual purpose of the
    class is to provide a query-like interface to the source files.
    """

    def __init__(self):
        self._sources = set()

    def __len__(self):
        """Get the number of candidates."""
        # see https://stackoverflow.com/a/15114062
        return len(self._sources)

    def add(self, new_source):
        """Add a new ``ImageCandidate``."""
        if isinstance(new_source, ImageCandidate):
            return self._sources.add(new_source)
        return NotImplemented

    def get_by_filename(self, filename):
        """Return an ``ImageCandidate`` by the associated filename."""
        # see https://stackoverflow.com/a/22693614
        #
        # The list comprehension *should* return just a single element, which
        # is returned with ``next(iter(...))``. If the comprehension is empty,
        # ``None`` is returned.
        return next(
            iter({item for item in self._sources if item.src_path == filename}), None
        )

    def get_fallback(self, fileformat):  # noqa D102
        return self.get_source_files(fileformat=fileformat)[0]

    def get_source_files(self, fileformat=None, min_width=0):  # noqa D102
        if fileformat is None:
            source_files = list(
                {item for item in self._sources if (item.width >= min_width)}
            )
        else:
            source_files = list(
                {
                    item
                    for item in self._sources
                    if (
                        item.width >= min_width and item.mime == EXT_TO_MIME[fileformat]
                    )
                }
            )
        source_files.sort(key=lambda can: can.width)

        return source_files


class ResponsiveImageCollector(EnvironmentCollector):
    """Collect responsive image source files in the build environment.

    This is heavily based on Sphinx's built-in ``ImageCollector``, see
    ``sphinx.environment.collectors.asset.ImageCollector``.
    """

    def clear_doc(self, app, env, docname):  # noqa D102
        logger.debug("ResponsiveImageCollector.clear_doc()")
        env.resp_img.purge_doc(docname)

    def merge_other(self, app, env, docnames, other):  # noqa D102
        logger.debug("ResponsiveImageCollector.merge_other()")
        env.resp_img.merge_other(docnames, other.resp_img)

    def process_doc(self, app, doctree):  # noqa D102
        logger.debug("ResponsiveImageCollector.process_doc()")


class respimage(nodes.General, nodes.Inline, nodes.Element):
    """The (custom) node to be inserted into the doctree.

    General (empty class):
    https://github.com/docutils/docutils/blob/173189b4c1c095a43c9388f4edd9bf1ff5d5b49d/docutils/docutils/nodes.py#L1194
    Inline (empty class):
    https://github.com/docutils/docutils/blob/173189b4c1c095a43c9388f4edd9bf1ff5d5b49d/docutils/docutils/nodes.py#L1217
    Element:
    https://github.com/docutils/docutils/blob/173189b4c1c095a43c9388f4edd9bf1ff5d5b49d/docutils/docutils/nodes.py#L439

    FIXME: Most likely this class is not required to provide custom functions,
           beside ``astext()``. Work will be done in the corresponding
           directive ``RespImage``.
    """

    # Overridden, as recommended in ``nodes.Element`` docstring.
    local_attributes = ()

    # def __init__(self, *args, **kwargs):
        # super().__init__(*args, **kwargs)

        # print("[DEBUG] attributes: {}".format(self.attributes))
        # print("[DEBUG] rawsource: {}".format(self.rawsource))

    def astext(self):
        """Return this node as text.

        Relying on the specified ``alt`` attribute.
        """
        return self.get("alt", "")


def visit_respimage_node(self, node):
    """Process the (custom) ``respimage`` node.

    This function provides the actual HTML output.

    See the corresponding call to ``add_node()`` in ``setup()``.
    """
    logger.debug("visit_respimage_node()")


def depart_respimage_node(self, node):
    """Process the (custom) ``respimage`` node.

    See the corresponding call to ``add_node()`` in ``setup()``.
    """
    logger.debug("depart_respimage_node()")


class RespImage(SphinxDirective):
    """Custom directive for *Responsive Images*."""

    # The directive requires exactly **one** argument, the path/filename of
    # the (responsive) image to be included.
    #
    # Please note: The existence of the specified image is not really required,
    # because the responsive ``<picture>`` element provides ``srcset`` elements
    # providing the different file formats and sizes.
    required_arguments = 1
    # This is already the default, but make it explicit!
    optional_arguments = 0
    # The argument may contain whitespaces. This allows the path/filename to
    # span multiple lines.
    final_argument_whitespace = True  # from docutil's Image directive

    # fom docutil's Image directive
    option_spec = {
        "alt": directives.unchanged,
        "height": directives.length_or_unitless,
        "width": directives.length_or_percentage_or_unitless,
        "class": directives.class_option,
    }

    def run(self):
        """Process the directive."""
        logger.debug("RespImage.run()")


def add_resp_img_to_env(app):
    """Add a ``FilenameUniqDict`` to the build environment.

    Sphinx original implementation to collect the required image files from
    ``:image:`` directives relies on this type of dict in the enviroment,
    accessible by ``env.images``, which is internall created in every
    environment.

    This is mimiced here for the responsive image files under the key
    ``env.resp_img``.
    """
    app.env.resp_img = FilenameUniqDict()


def setup(app):
    """Register the extension with Sphinx.

    This function is required by Sphinx's extension interface, see
    https://www.sphinx-doc.org/en/master/development/tutorials/helloworld.html#writing-the-extension
    for reference.
    """
    # This configuration value maps image formats to their corresponding output
    # priorities (lower priority value = higher logical priority).
    #
    # The items in the list are of type ``tuple``, where the first element is
    # the target format, specified **including** a leading full stop (``.``).
    # The second item is the corresponding **output priority**.
    #
    # Note: The **output priority** also controls the order of processing the
    #       source files. While *lower* priority values mean a higher logical
    #       output priority, *higher* priority values lead to earlier
    #       processing for candidates.
    app.add_config_value(
        "responsive_images_formats",
        [
            (".avif", 1),
            (".jpg", 99),
            (".png", 98),
            (".webp", 2),
        ],
        "env",
    )

    # This configuration value provides a list of filename suffixes to indicate
    # the image dimensions.
    #
    # It is recommended to provide a *semantic name*. The values are appended
    # to the *stem* of the given source filename to construct the filenames of
    # the responsive variants with different sizes.
    app.add_config_value(
        "responsive_images_size_suffixes",
        ["-tiny", "-small", "-medium", "-big", "-large", "-xlarge"],
        "env",
    )

    # This configuration value maps *layout breakpoints* to minimum required
    # *image widths*.
    #
    # During processing, a default *non-breakpoint* ``(0, 0)`` is provided
    # automatically.
    #
    # TODO: Provide a sane default value, probably even an empty list.
    # TODO: Add documentation about the required format (tuple, including
    #       meaning of the elements)
    app.add_config_value(
        "responsive_images_layout_breakpoints",
        [(777, 444), (500, 222)],
        "env",
    )

    # Add a custom node for responsive images.
    #
    # Note: As of now, this only includes corresponding generator functions
    #       (``visit_``, ``depart_``) for HTML writers.
    app.add_node(respimage, html=(visit_respimage_node, depart_respimage_node))

    # Add a custom directive for responsive images.
    app.add_directive("resp_image", RespImage)

    app.connect("builder-inited", add_resp_img_to_env)

    app.add_env_collector(ResponsiveImageCollector)

    return {
        "version": "0.0.1",
        "env-version": "1",
        "parallel_read_safe": True,
        "parallel_write_safe": True,
    }
```

**Comment by Mischback** [comment]
There is a follow-up issue (#61)!

**Comment by Mischback** [comment]
Closed with 6012879c97aec45a686a819410ab3ad4768b646e / #56
